gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java

```diff
@@ -0,0 +1,554 @@
+    int priority = initialPriority;
+    if (this.tableRegistrationStep.isPresent()) {
+      copyEntities.add(new PrePublishStep(fileSet, Maps.<String, Object>newHashMap(), this.tableRegistrationStep.get(),
+          priority++));
+    }
+    return priority;
+  }
+
+  private List<CopyEntity> getCopyEntitiesForUnpartitionedTable() throws IOException {
+
+    int stepPriority = 0;
+    String fileSet = this.dataset.table.getTableName();
+    List<CopyEntity> copyEntities = Lists.newArrayList();
+
+    stepPriority = addSharedSteps(copyEntities, fileSet, stepPriority);
+
+    InputFormat<?, ?> inputFormat = HiveDataset.getInputFormat(this.dataset.table.getSd());
+
+    Collection<Path> sourcePaths = HiveDataset.getPaths(inputFormat, this.dataset.table.getSd().getLocation());
+    Collection<Path> targetExistingPaths = this.existingTargetTable.isPresent() ?
+        HiveDataset.getPaths(inputFormat, this.existingTargetTable.get().getSd().getLocation()) : Sets.<Path>newHashSet();
+
+    DiffPathSet diffPathSet = diffSourceAndTargetPaths(sourcePaths, targetExistingPaths,
+        Optional.<Partition>absent(), targetFs);
+
+    DeleteFileCommitStep deleteStep = DeleteFileCommitStep.fromPaths(targetFs, diffPathSet.pathsToDelete,
+        this.dataset.properties);
+    copyEntities.add(new PrePublishStep(fileSet, Maps.<String, Object>newHashMap(), deleteStep, stepPriority++));
+
+    for (CopyableFile.Builder builder : getCopyableFilesFromPaths(diffPathSet.filesToCopy,
+        configuration, Optional.<Partition>absent())) {
+      copyEntities.add(builder.build());
+    }
+
+    return copyEntities;
+  }
+
+  private DiffPathSet diffSourceAndTargetPaths(Collection<Path> sourcePaths, Collection<Path> targetExistingPaths,
+      Optional<Partition> partition, FileSystem targetFs) throws IOException {
+    DiffPathSet.DiffPathSetBuilder builder = DiffPathSet.builder();
+
+    Set<Path> targetPathSet = Sets.newHashSet(targetExistingPaths);
+
+    for (Path sourcePath : sourcePaths) {
+      FileStatus status = this.dataset.fs.getFileStatus(sourcePath);
+      Path targetPath = targetFs.makeQualified(getTargetPath(status, this.targetFs, partition));
+      if (targetPathSet.contains(targetPath)) {
+        FileStatus targetStatus = targetFs.getFileStatus(targetPath);
+        if (targetStatus.getLen() != status.getLen() || targetStatus.getModificationTime() < status.getModificationTime()) {
+          // We must replace target file
+          builder.deleteFile(targetPath);
+          builder.copyFile(status);
+        }
+        targetPathSet.remove(targetPath);
+      } else {
+        builder.copyFile(status);
+      }
+    }
+
+    // All leftover paths in target are not in source, should be deleted.
+    for (Path deletePath : targetPathSet) {
+      builder.deleteFile(deletePath);
+    }
+
+    return builder.build();
+  }
+
+  private void checkTableCompatibility(Table sourceTable, Table targetTable) throws IOException {
```

Comment:
Because there is more information in an `IOException`, including a message with the precise reason and a stack trace.